cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -115,49 +117,61 @@ protected void run() throws Exception {
 
     ApplicationSpecification appSpec = program.getApplicationSpecification();
 
-    Iterator<ScheduleProgramInfo> iterator = workflowSpec.getActions().iterator();
+    Iterator<WorkflowNode> iterator = workflowSpec.getNodes().iterator();
     int step = 0;
     while (running && iterator.hasNext()) {
       WorkflowActionSpecification actionSpec;
-      ScheduleProgramInfo actionInfo = iterator.next();
-      switch (actionInfo.getProgramType()) {
-        case MAPREDUCE:
-          MapReduceSpecification mapReduceSpec = appSpec.getMapReduce().get(actionInfo.getProgramName());
-          actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(mapReduceSpec.getName(),
-                                                                                        mapReduceSpec.getName(),
-                                                 SchedulableProgramType.MAPREDUCE));
+      WorkflowNode node = iterator.next();
+      switch (node.getType()) {
+        case ACTION:
+          WorkflowActionNode actionNode = (WorkflowActionNode) node;
```

Comment:
There are some related changes already done in the PR #1386. Since that PR is touching this part of code majorly, will make sure to make these changes in that PR.